app/v2/controllers/GuaranteeBalanceController.scala

```diff
@@ -61,24 +70,40 @@ class GuaranteeBalanceController @Inject() (
     authenticate().io(parse.json) {
       implicit request =>
         (for {
-          _                <- validateAcceptHeader
-          _                <- lockService.lock(grn, request.internalId).asPresentation
-          parsed           <- parseJson(request.body)
-          _                <- validationService.validate(parsed).asPresentation
-          internalResponse <- routerService.request(grn, parsed).asPresentation
+          _      <- validateAcceptHeader
+          parsed <- parseJson(request.body)
+          auditInfo = AuditInfo(parsed, grn, request.internalId)
+          _                <- lockService.lock(grn, request.internalId).asPresentation(auditInfo, auditService)
+          _                <- validationService.validate(parsed).asPresentation(auditInfo, auditService)
+          internalResponse <- routerService.request(grn, parsed).asPresentation(auditInfo, auditService)
           hateoas          <- EitherT.right[PresentationError](HateoasResponse(grn, internalResponse))
+          _ = auditService.balanceRequestSucceeded(auditInfo, internalResponse.balance)
         } yield hateoas).fold(
-          presentationError => Status(presentationError.code.statusCode)(Json.toJson(presentationError)),
+          presentationError => requestFailed(request, grn, presentationError),
           result => Ok(result)
         )
     }
 
+  private lazy val accessCodeErrorMsg   = "The access code was not supplied."
+  private lazy val acceptHeaderErrorMsg = "The accept header must be set to application/vnd.hmrc.2.0+json to use this resource."
+
+  private def requestFailed(request: AuthenticatedRequest[JsValue], guaranteeReferenceNumber: GuaranteeReferenceNumber, presentationError: PresentationError)(
+    implicit hc: HeaderCarrier
+  ): Result = {
+    presentationError.message match {
+      case `accessCodeErrorMsg` | `acceptHeaderErrorMsg` =>
+        auditService.invalidPayloadBalanceRequest(request, guaranteeReferenceNumber)
+      case _ => // don't send second audit event - already sent during presentation error determination.
+    }
+    Status(presentationError.code.statusCode)(Json.toJson(presentationError))
+  }
```

Comment:
Can't we just do the auditing in `validateAcceptHeader` and `parseJson` at the point we know we have a left?

Reply:
Yeah, good point.